### PR TITLE
RHCEPHQE-18632 | [RADOS] Enable multiple upgrades using 'rhcs-version' and 'release' in test config

### PR DESCRIPTION
**Pull request by harshkumarRH**
# Description

Jira tracker: [RHCEPHQE-18632](https://issues.redhat.com/browse/RHCEPHQE-18632)

Code changes in this PR enable user to trigger multiple upgrade in a single test suite by providing `rhcs-version` and `release` as input configs
Currently, this support has only been added in `tests/rados/test_upgrade_warn.py` test under RADOS but can be extended to the general/common test used for upgrading a cluster by most FGs.

Multi-step upgrade path is essential to verify customer bugs that were encountered during upgrades and whose fix was provided in later releases. This also provides us the ability to add more complex customer centric test workflows to increase coverage and uncover edge case bugs.

Sample test:
```
  - test:
      name: Upgrade cluster to 7.1z2 ceph version
      desc: Upgrade cluster to 7.1z2 version
      module: test_upgrade_warn.py
      polarion-id: CEPH-83574934
      config:
        command: start
        service: upgrade
        base_cmd_args:
          verbose: true
        args:
          rhcs-version: 7.1
          release: z2
      abort-on-fail: true
```

Logs:
Bootstrap cluster with RHCS 7.1z1 -> Upgrade to RHCS 7.1z2 -> Upgrade to RHCS 8.0 nightly latest

<details>

<summary>click to expand checklist</summary>

- [ ] Create a test case in Polarion reviewed and approved.
- [ ] Create a design/automation approach doc. Optional for tests with similar tests already automated.
- [ ] Review the automation design
- [ ] Implement the test script and perform test runs
- [ ] Submit PR for code review and approve
- [ ] Update Polarion Test with Automation script details and update automation fields
- [ ] If automation is part of Close loop, update BZ flag qe-test_coverage “+” and link Polarion test
</details>

### CODE REVIEW

File: tests/rados/test_upgrade_warn.py [hunk 1]
```diff
@@ -82,7 +110,9 @@ def run(ceph_cluster, **kw):
         log.info(f"Current version on the cluster : {ceph_version}")
 
         # Start Upgrade
-        config.update({"args": {"image": "latest"}})
+        if not (_rhcs_release and _rhcs_version):
+            log.info("WE ARE GETTING IN THIS BLOCK!!!")
```
Comment (pdhiran) [inline tests/rados/test_upgrade_warn.py:114]: update log

File: tests/rados/test_upgrade_warn.py [hunk 2]
```diff
@@ -70,9 +74,33 @@ def run(ceph_cluster, **kw):
...
+
+            # The cluster object is configured so that the values are persistent till
+            # an upgrade occurs. This enables us to execute the test in the right
+            # context.
+            config["base_url"] = _details[0]
+            config["container_image"] = f"{_details[1]}/{_details[2]}:{_details[3]}"
+            ceph_cluster.rhcs_version = _rhcs_version or rhbuild
+
+        # Remove existing repos
+        rm_repo_cmd = "find /etc/yum.repos.d/ -type f ! -name hashicorp.repo ! -name redhat.repo -delete"
```
Comment (pdhiran) [inline tests/rados/test_upgrade_warn.py:95]: `yum clean all`

**Review by pdhiran — COMMENTED** [review]
Looks good with minor changes

**Comment by pdhiran** [comment]
@vamahaja , can you please review the PR?

File: ceph/ceph_admin/upgrade.py [hunk 3]
```diff
@@ -52,7 +52,10 @@ def start_upgrade(self: OrchProtocol, config: Dict):
 
         args = config.get("args")
         if "image" in args:
-            cmd.append(f"--image {self.config.get('container_image')}")
+            _image = self.config.get("container_image")
```
Comment (vamahaja) [inline ceph/ceph_admin/upgrade.py:55]: Why image is passed from two separate places ? What's the difference between `config` and `self.config`? What are different scenarios and which one will use which parameter?

Comment (harshkumarRH) [reply ceph/ceph_admin/upgrade.py:74]: Rationale -
`self.config` is the kwargs dict that is prepared in run.py and is sent as an argument when any test is triggered.

`self.config` contains the ceph image and repo url detail as determined by the input to <`--rhbuild`> in run.py
In order to retain the global config, changes are made to local config that is passed to this method.
Changes made in local config dict:
- modify `container_image` entry to desired ceph image url
- modify `base_url` entry to desired ceph repo url
When this method is called and input config dict contains `{"args": {"image": "latest"}}`, the upgrade goes ahead with
the original test config dict(`self.config`)
If `args` entry in input config dict does not contain `image` key, it is understood that user must have set a custom `container_image` and `base_url` in local config dict, and those are picked to proceed with the upgrade process.

Even if user does not modify the local config dict, upgrade will still succeed as the config dict would still contain the unmodified `container_image` and `'base_url`

However, in hindsight, I reviewed the code the again and realized that we may not actually need to go through these hoops to enable multi level upgrade. It seems the ability to initiate multi level upgrade was kind-of already present.

The config dict prepared by run.py is sent to each test during execution, this dict remains unchanged and therefore needs no retention.
In other words, user should be able to modify the config dict with local changes, i.e. replace the `base_url` and `container_image` with custom values without maintaining two set of config dictionaries.
Once config dict has been modified, user just needs to initialize a new class object for Orch(`ceph/ceph_admin/orch.py`) class with the updated config to proceed with the upgrade.

File: tests/rados/test_upgrade_warn.py [hunk 4]
```diff
@@ -70,9 +74,47 @@ def run(ceph_cluster, **kw):
 
     log.debug("Starting upgrade")
     try:
+        config.update({"args": {"image": "latest"}})
+
+        # Support installation of the baseline cluster whose version is not available in
+        # CDN. This is primarily used for an upgrade scenario. This support is currently
+        # available only for RH network.
+        _rhcs_version = args.get("rhcs-version", None)
+        _rhcs_release = args.get("release", None)
```
Comment (pdhiran) [inline tests/rados/test_upgrade_warn.py:83]: please dump the version infos

**Review by vamahaja — APPROVED** [review]

**Review by pdhiran — APPROVED** [review]
looks good

**Review by AmarnatReddy — APPROVED** [review]
lgtm

**Review by pranavprakash20 — APPROVED** [review]
LGTM
